### PR TITLE
[Update](deps) add clucene deps for doris inverted index

### PR DESCRIPTION
**Pull request by airborne12**
# Proposed changes

As part of [Inverted Index DSIP](https://cwiki.apache.org/confluence/display/DORIS/DSIP-023%3A+Add+inverted+index+for+full+text+search) steps, we'd like to contribute our inverted index implementations step by step.
First of all we need to introduce clucene to doris thirdparty libs, because inverted index implementations are based on lucence API and index file format, also we add our features and performance improvements base on clucene, so we need to maintain the repo ourselves

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by morningman — DISMISSED** [review]
LGTM

File: thirdparty/build-thirdparty.sh [hunk 1]
```diff
@@ -1545,12 +1545,43 @@ build_concurrentqueue() {
...
 }
 
+#clucene
+build_clucene() {
+    if [[ -z ${USE_AVX2} ]]; then
+        USE_AVX2=1
+    fi
+    if [[ -z ${BUILD_TYPE} ]]; then
+        BUILD_TYPE=Release
+    fi
```
Comment (adonis0147) [inline thirdparty/build-thirdparty.sh:1555]: This block can be omitted. Just pass `-DCMAKE_BUILD_TYPE=RelWithDebInfo` to CMake.

File: thirdparty/build-thirdparty.sh [hunk 2]
```diff
@@ -1545,12 +1545,43 @@ build_concurrentqueue() {
...
+        CPPFLAGS="-fno-omit-frame-pointer -g -Wno-narrowing" \
+            CXXFLAGS="-fno-omit-frame-pointer -g -Wno-narrowing" \
+            ${CMAKE_CMD} -G "${GENERATOR}" -DCMAKE_INSTALL_PREFIX="${TP_INSTALL_DIR}" -DBUILD_STATIC_LIBRARIES=ON \
+            -DUSE_AVX2="${USE_AVX2}" -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" -DBUILD_CONTRIBS_LIB=ON ..
+        ${BUILD_SYSTEM} -j "${PARALLEL}" && ${BUILD_SYSTEM} install
+    elif [[ "${CC}" == *clang ]]; then
+        CPPFLAGS="-fno-omit-frame-pointer -g -Wno-c++11-narrowing" \
+            CXXFLAGS="-fno-omit-frame-pointer -g -Wno-c++11-narrowing" \
+            ${CMAKE_CMD} -G "${GENERATOR}" -DCMAKE_INSTALL_PREFIX="${TP_INSTALL_DIR}" -DBUILD_STATIC_LIBRARIES=ON \
+            -DUSE_AVX2="${USE_AVX2}" -DCMAKE_BUILD_TYPE="${BUILD_TYPE}" -DBUILD_CONTRIBS_LIB=ON ..
```
Comment (adonis0147) [inline thirdparty/build-thirdparty.sh:1571]: Looks verbose. The following common part can be separated like

```shell
${CMAKE_CMD} -G "${GENERATOR}" -DCMAKE_INSTALL_PREFIX="${TP_INSTALL_DIR}" -DBUILD_STATIC_LIBRARIES=ON \
            -DCMAKE_CXX_FLAGS="-fno-omit-frame-pointer ${warning_narrowing}"
            -DUSE_AVX2="${USE_AVX2}" -DCMAKE_BUILD_TYPE="RelWithDebInfo" -DBUILD_CONTRIBS_LIB=ON ..
${BUILD_SYSTEM} -j "${PARALLEL}"
${BUILD_SYSTEM} install
```

File: thirdparty/build-thirdparty.sh [hunk 3]
```diff
@@ -1545,12 +1545,43 @@ build_concurrentqueue() {
...
+        BUILD_TYPE=Release
+    fi
+    check_if_source_exist "${CLUCENE_SOURCE}"
+    cd "${TP_SOURCE_DIR}/${CLUCENE_SOURCE}"
+    mkdir -p "${BUILD_DIR}" && cd "${BUILD_DIR}"
+    rm -rf CMakeCache.txt CMakeFiles/
+
+    if [[ "${CC}" == *gcc ]]; then
+        CPPFLAGS="-fno-omit-frame-pointer -g -Wno-narrowing" \
+            CXXFLAGS="-fno-omit-frame-pointer -g -Wno-narrowing" \
```
Comment (adonis0147) [inline thirdparty/build-thirdparty.sh:1563]: Using `CXX_FLAGS` is enough.

Comment (adonis0147) [reply thirdparty/build-thirdparty.sh:1563]: It is better to use a variable like `warning_narrowing`. See [L144](https://github.com/apache/doris/blob/master/thirdparty/build-thirdparty.sh#L144)

**Review by adonis0147 — CHANGES_REQUESTED** [review]

File: thirdparty/build-thirdparty.sh [hunk 4]
```diff
@@ -1545,12 +1545,43 @@ build_concurrentqueue() {
     cp ./*.h "${TP_INSTALL_DIR}/include/"
 }
 
+#clucene
+build_clucene() {
+    if [[ -z ${USE_AVX2} ]]; then
+        USE_AVX2=1
+    fi
```
Comment (adonis0147) [inline thirdparty/build-thirdparty.sh:1552]: Is the variable needed in CLucene?

File: thirdparty/build-thirdparty.sh (hunk 1 above)
Comment (airborne12) [reply thirdparty/build-thirdparty.sh:1555]: Only with RelWithDebInfo flags is not enough, CLucene also need to compile ASAN edition,

File: thirdparty/build-thirdparty.sh (hunk 4 above)
Comment (airborne12) [reply thirdparty/build-thirdparty.sh:1552]: Yes, we use TurboPFor for integer compression which need AVX2 compile flags to seperate different building arch

Comment (adonis0147) [reply thirdparty/build-thirdparty.sh:1552]: Okey.

File: thirdparty/build-thirdparty.sh (hunk 1 above)
Comment (adonis0147) [reply thirdparty/build-thirdparty.sh:1555]: > Only with RelWithDebInfo flags is not enough, CLucene also need to compile ASAN edition,

It is a third-party library, we can just use it. Why do we need ASAN build?

Comment (airborne12) [reply thirdparty/build-thirdparty.sh:1555]: It's not just a third-party lib, we added a lot features into it, sometimes we need ASAN to debug problems

Comment (adonis0147) [reply thirdparty/build-thirdparty.sh:1555]: > It's not just a third-party lib, we added a lot features into it, sometimes we need ASAN to debug problems

Okey.

File: thirdparty/build-thirdparty.sh (hunk 3 above)
Comment (airborne12) [reply thirdparty/build-thirdparty.sh]: done

File: thirdparty/build-thirdparty.sh (hunk 2 above)
Comment (airborne12) [reply thirdparty/build-thirdparty.sh]: done, good suggestion, thanks

File: thirdparty/build-thirdparty.sh (hunk 3 above)
Comment (adonis0147) [reply thirdparty/build-thirdparty.sh]: By the way, CMake don't provide `ASAN` build type by default. You would better add `-g` to these variables ([CXX_FLAGS_ASAN](https://github.com/apache/doris-thirdparty/blob/clucene/CMakeLists.txt#L120) and [CXX_FLAGS_LSAN](https://github.com/apache/doris-thirdparty/blob/clucene/CMakeLists.txt#L121)) in `CMakeLists.txt` of [clucene](https://github.com/apache/doris-thirdparty/tree/clucene).

Examples:
```CMakeLists.txt
if (__COMPILER_CLANG)
    SET(CXX_FLAGS_ASAN "${CXX_GCC_FLAGS} -O0 -g -fsanitize=address -DADDRESS_SANITIZER")
    SET(CXX_FLAGS_LSAN "${CXX_GCC_FLAGS} -O0 -g -fsanitize=leak -DLEAK_SANITIZER")
else ()
    SET(CXX_FLAGS_ASAN "${CXX_GCC_FLAGS} -O0 -g -fsanitize=address -DADDRESS_SANITIZER -static-libasan")
    SET(CXX_FLAGS_LSAN "${CXX_GCC_FLAGS} -O0 -g -fsanitize=leak -DLEAK_SANITIZER -static-liblsan")
endif ()
```

Comment (airborne12) [reply thirdparty/build-thirdparty.sh]: Already done that in CLucene, thanks a lot~

**Review by adonis0147 — DISMISSED** [review]
LGTM

**Review by qidaye — DISMISSED** [review]
LGTM

**Comment by adonis0147** [comment]
It seems that we need an extra works to port the library CLucene to macOS.

**Comment by airborne12** [comment]
> 

Yeah, also noticed that

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.14 seconds
 load time: 492 seconds
 storage size: 17122179885 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230111190005_clickbench_pr_78059.html

**Review by morningman — APPROVED** [review]
LGTM

**Review by adonis0147 — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**